mpc123.1
.TH mpc123 1 "January 2006"
.SH NAME
mpc123 \- your handy Musepack audio player
.SH SYNOPSIS
.B mpc123 [options] file(s)
.SH DESCRIPTION
mpc123 is a command\-line player for the Musepack audio compression
format. mpc123 reads mpc audio files and decodes them to the devices
specified on the command line.

During playback, you can pass to the next playlist element (or a random one
if one of the \-z and \-Z options were used, see below) by sending SIGINT
to the process, thus usually pressing ^C will act as a "next-button" for the
player.

.SH OPTIONS
.IP "\fB\-\-gain N\fP, \fB\-g N\fP" 4
Set gain (audio volume) to \fBN\fP (0\-100 inclusive, default 100,
0 mutes sound)

.IP "\fB\-o driver\fP" 4
Set output devicetype to \fBdriver\fP; possible drivers include:
.IP "  \fBoss\fP" 4
Linux Open Sound System
.IP "  \fBalsa\fP" 4
Advanced Linux Sound Architecture
.IP "  \fBalsa09\fP" 4
ALSA version 0.9 and newer (you should really be using this)
.IP "  \fBesd\fP" 4
Enlightenment sound daemon
.IP "  \fBarts\fP" 4
Analog Real Time System (kde sound daemon)
.IP "  \fBnull\fP" 4
Debugging output target

.IP "\fB\-\-audiodevice dev\fP, \fB\-a dev\fP" 4
Use \fBdev\fP for audio\-out; If not specified, the program tries to pick
some sane defaults, based on the used audio output driver. Common devices
include:
.IP "  \fB/dev/dsp\fP or \fB/dev/dsp1\fP" 4
for \fBoss\fP output
.IP "  \fBdefault\fP or \fBhw:0\fP" 4
for \fBalsa09\fP output

.IP "\fB\-\-au filename.au\fP, \fB\-u filename.au\fP"
Use au file \fBfilename.au\fP for output

.IP "\fB\-\-cdr filename.cdr\fP, \fB\-c filename.cdr\fP"
Use raw file \fBfilename.cdr\fP for output (this can be used directly
with cdrecord's \-audio option)

.IP "\fB\-\-wav filename.wav\fP, \fB\-w filename.wav\fP"
Use wave file \fBfilename.wav\fP for output

.IP "\fB\-\-list file\fP, \fB\-@ file\fP"
Use playlist \fBfile\fP as list of Musepack files; The playlist format is
simple: one file per line.
You can easily generate a playlist for \fBmpc123\fP with the find(1)
command:
.IP "    find /path/to/musicroot \-iname \\\\*.mpc \-fprint \fBfile\fP" 4

.IP "\fB\-\-random\fP, \fB\-Z\fP"
Play files randomly until interrupted

.IP "\fB\-\-shuffle\fP, \fB\-z\fP"
Shuffle list of files before playing

.IP "\fB\-\-verbose\fP, \fB\-v\fP"
Increase verbosity (default verbosity is 0), the more \-v, the more
verbose mpc123 becomes

.IP "\fB\-\-quiet\fP, \fB\-q\fP"
Reset verbosity to 0 (no title or boilerplate); this is the default

.IP "\fB\-\-help\fP, \fB\-h\fP"
Print the help screen, with some brief usage information

.SH AUTHOR
.na
.nf
This manual page was written by Daniele Sempione <scrows at oziosi.org>; Fernando Vezzosi <github-mpc123 at repnz.net> made some edits


ChangeLog
# do not edit -- automatically generated by arch changelog
# arch-tag: automatic-ChangeLog--bucciarati@users.sourceforge.net--2005--mpc123/mpc123--main--0.1
#

2006-08-02 11:26:35 GMT	Fernando Vezzosi <fvezzosi@masobit.net>	patch-33

    Summary:
      minor tweaks for release 0.2
    Revision:
      mpc123--main--0.1--patch-33

     * added info about the mailing list
     * cleaned Makefile out of useless stuff
     * updated TODO to reflect the actual TODO :)
    

    modified files:
     AUTHORS ChangeLog Makefile TODO


2006-08-02 10:20:20 GMT	Daniele Sempione <scrows@oziosi.org>	patch-32

    Summary:
      major/minor fix
    Revision:
      mpc123--main--0.1--patch-32

    * changed major/minor for the release :)
    

    modified files:
     ChangeLog Makefile mpc123.h


2006-07-31 18:32:42 GMT	Fernando Vezzosi <fvezzosi@masobit.net>	patch-31

    Summary:
      exit if two sigints are sent in a short time window
    Revision:
      mpc123--main--0.1--patch-31

     * now it exits if two ^C's are pressed in a rapid succession
     * fixed some bugs in the implementation ;)
    

    modified files:
     ChangeLog mpc123.c player.c playlist.c signals.c


2006-07-31 16:59:58 GMT	Daniele Sempione <scrows@oziosi.org>	patch-30

    Summary:
      sighandler is thread-safe
    Revision:
      mpc123--main--0.1--patch-30

    * now SIGINT is thread-safe
    * added support for terminating program via double SIGINT.
      now set to 0.7 secs, let's test it .. (still needed buccia's cleanup
      code, right now it only sets the flag)
    

    modified files:
     ChangeLog signals.c


2006-07-31 15:42:24 GMT	Fernando Vezzosi <fvezzosi@masobit.net>	patch-29

    Summary:
      Updated documentation (README and manpage)
    Revision:
      mpc123--main--0.1--patch-29

     * added info about esd
     * updated info about program behavior (SIGINT handling
       and choosing of default ao device)
    

    modified files:
     ChangeLog README mpc123.1


2006-07-31 15:11:09 GMT	Daniele Sempione <scrows@oziosi.org>	patch-28

    Summary:
      added 2006 to some copyrights
    Revision:
      mpc123--main--0.1--patch-28

    * ...
    

    modified files:
     ChangeLog Makefile README


2006-07-31 14:40:01 GMT	Daniele Sempione <scrows@oziosi.org>	patch-27

    Summary:
      SIGINT = next song
    Revision:
      mpc123--main--0.1--patch-27

    * now this feature works!! :))
    * ported signal handling to sigaction
    

    new files:
     .arch-ids/signals.c.id signals.c

    modified files:
     ChangeLog mpc123.c mpc123.h


2006-07-31 13:54:37 GMT	Fernando Vezzosi <fvezzosi@masobit.net>	patch-26

    Summary:
      reverted my stupid edits
    Revision:
      mpc123--main--0.1--patch-26

     * i will not mess with a thing i shouldn't have been messing with again
     * i will not use wrong string format specifiers again
    
     sorry world
    

    modified files:
     ChangeLog mpc123.h reader_file.c


2006-07-31 13:47:38 GMT	Fernando Vezzosi <fvezzosi@masobit.net>	patch-25

    Summary:
      fixed "next-song" bug, added vim modelines
    Revision:
      mpc123--main--0.1--patch-25

     * fixed a dumb bug of my previous commit, which made the player
       skip all playlist elements, when it should have skipped only
       the currently played one..
     * added the modeline "ft=c:tw=78:ts=2:et:cin" to all source files
    

    modified files:
     ChangeLog ao.c mpc123.c mpc123.h player.c playlist.c reader.c
     reader_file.c shuffle.c


2006-07-31 13:13:19 GMT	Fernando Vezzosi <fvezzosi@masobit.net>	patch-24

    Summary:
      Added macro functions for flag handling
    Revision:
      mpc123--main--0.1--patch-24

     * added flags global variable
     * added macros for flag handling
     * player loop stops if correct flags are set
    

    modified files:
     ChangeLog mpc123.h player.c


2006-07-30 10:17:38 GMT	Daniele Sempione <scrows@oziosi.org>	patch-23

    Summary:
      Makefile comment
    Revision:
      mpc123--main--0.1--patch-23

    * disables the -lefence linking option I forgot in the Makefile
    

    modified files:
     ChangeLog Makefile


2006-07-30 02:19:04 GMT	Fernando Vezzosi <fvezzosi@masobit.net>	patch-22

    Summary:
      Fixes in view of new release
    Revision:
      mpc123--main--0.1--patch-22

     * removed generation of locales in Makefile
     * better support for all sound systems, including working support for:
       - alsa09
       - esd
     * decent libao defaults are now set at runtime
     * improved man page with content and appearance
     * improved debugging output
     * cosmetic changes
    

    modified files:
     ChangeLog Makefile ao.c mpc123.1 mpc123.c mpc123.h


2006-07-29 13:38:47 GMT	Daniele Sempione <scrows@oziosi.org>	patch-21

    Summary:
      removed bugged l10n support
    Revision:
      mpc123--main--0.1--patch-21

    * l10n was too buggy, we'll re-add in the future
    * updated copyrights' year
    

    modified files:
     ChangeLog Makefile ao.c mpc123.c mpc123.h player.c playlist.c
     reader.c reader_file.c shuffle.c


2006-07-28 13:33:57 GMT	Fernando Vezzosi <fvezzosi@masobit.net>	patch-20

    Summary:
      fixed segfault when using inappropriate commandline flags
    Revision:
      mpc123--main--0.1--patch-20

     * a crappy bug that lead to sigsegv upon invocation
    

    modified files:
     ChangeLog mpc123.c


2006-07-28 13:28:17 GMT	Fernando Vezzosi <fvezzosi@masobit.net>	patch-19

    Summary:
      
    Revision:
      mpc123--main--0.1--patch-19

    
    

    modified files:
     ChangeLog Makefile mpc123.h


2006-06-06 18:39:42 GMT	Daniele Sempione <scrows@oziosi.org>	patch-18

    Summary:
      added '\n' to some strings
    Revision:
      mpc123--main--0.1--patch-18

    * ...
    

    modified files:
     ChangeLog ao.c mpc123.c mpc123.h player.c reader.c


2006-03-22 18:07:36 GMT	Fernando Vezzosi <fvezzosi@masobit.net>	patch-17

    Summary:

mpc123.c
/*
 *  mpc123 - Musepack Console audio player
 *  Copyright (C) 2005-2008 Fernando Vezzosi <github-mpc123 at repnz.net>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software Foundation,
 *  Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

/* is this portable ? */
#define _GNU_SOURCE

#include <getopt.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <stdarg.h>
#include <fcntl.h>
#include <locale.h>
#include <libintl.h>

#include "mpc123.h"

#ifdef DEBUG
#  include <mcheck.h>
#endif

#define ARG_YES 1
#define ARG_NO 0
#define ARG_MAYBE 2

/* (hopefully) sane defaults */
opts_t options={
  .shuffle=0,
  .random=0,
  .verbosity=0,
  .foutput=NULL,
  .playlist=NULL,
  .ao_driver=NULL,
  .ao_dev=NULL,
  .volume=1.0
};

static void version(){
  fprintf(stderr, _("mpc123 version %d.%d.%d\n"),
          VERS_MAJOR, VERS_MINOR, VERS_REV);
  fprintf(stderr, "Copyright %s\n", COPYRIGHT);
}

static void usage(const char *name){
  version();
  sayf(0, "\nUsage: %s [options] [file1 [file2 [..]]]\n\n"
          "Options supported:\n"
          "   --verbose or -v          Increase verbosity\n"
          "   --quiet or -q            Quiet mode (no title or boilerplate)\n"
          "   --gain N or -g N         Set gain (audio volume) to N (0-100)\n"
          "   -o dt                    Set output devicetype to dt\n"
          "   --audiodevice N or -a N  Use N for audio-out\n"
          "   --au N or -u N           Use au file N for output\n"
          "   --cdr N or -c N          Use raw file N for output\n"
          "   --wav N or -w N          Use wave file N for output\n"
          "   --list N or -@ N         Use playlist N as list of Musepack files\n"
          "   --random or -Z           Play files randomly until interrupted\n"
          "   --shuffle or -z          Shuffle list of files before playing\n"
          "   --help or -h             Print this help screen\n"
          "   --version or -V          Print mpc123 version string\n",
	name);
}

int main(int argc, char ** argv){
  int c=0;
  int ret=0;
  playlist_t pl;

  /*  mpc_bool_t mpc_ret=FALSE;*/
  struct option myOpt[] = {
    {"verbose", ARG_NO, 0, 'v'},
    {"quiet", ARG_NO, 0, 'q'},
    {"gain", ARG_YES, 0, 'g'},
    {"audiodevice", ARG_YES, 0, 'a'},
    {"au", ARG_YES, 0, 'u'},
    {"cdr", ARG_YES, 0, 'c'},
    {"wav", ARG_YES, 0, 'w'},
    {"list", ARG_YES, 0, '@'},
    {"random", ARG_NO, 0, 'Z'},
    {"shuffle", ARG_NO, 0, 'z'},
    {"help", ARG_NO, 0, 'h'},
    {"version", ARG_NO, 0, 'V'},
    {NULL, 0, 0, '\0'}
  };

  setlocale(LC_ALL, "");
  bindtextdomain(PACKAGE, LOCALEDIR);
  textdomain(PACKAGE);

#ifdef DEBUG
  /* see glibc-doc, chapter "Debugging Memory Allocation"
   * for the use of mcheck.h, mtrace(), MALLOC_TRACE and mtrace(1);
   */
  mtrace();
  debugf("mpc123 compiled with DEBUG=%d enabled", DEBUG);
  if(!getenv("MALLOC_TRACE")){
    debug("WARNING: you have not set a MALLOC_TRACE destination file");
  }else{
    debug("* * malloc_trace activated; to check mpc123 for memory leaks,");
    debugf("* * type the following command: \"mtrace %s %s\"",
           argv[0], getenv("MALLOC_TRACE"));
  }
#endif

  while ((c=getopt_long(argc, argv, _GETOPT_FLAGS, myOpt, NULL)) != -1){
    switch(c){
      case 'z':			/* shuffle the pl once */
        options.shuffle=1;
        break;
      case 'Z':			/* choose a random entry from pl each time */
        options.random=1;
        break;
      case 'v':			/* be logorroical */
        options.verbosity++;
        break;
      case 'q':			/* be asocial */
        options.verbosity=0;
        break;
      case '@':     /* playlist */
        options.playlist=optarg;
        break;
      case 'o':     /* output driver */
        options.ao_driver=optarg;
        break;
      case 'a':     /* libao device to use */
        options.ao_dev=optarg;
        break;
      case 'w':     /* output to .wav */
        options.ao_driver="wav";
        options.foutput=optarg;
        break;
      case 'u':     /* output to .au */
        options.ao_driver="au";
        options.foutput=optarg;
        break;
      case 'c':     /* output to raw pcm */
        options.ao_driver="raw";
        options.foutput=optarg;
      case 'g':     /* set gain */
        options.volume=atof(optarg)/100.0;
        if(options.volume<0.0 || options.volume>1.0){
          dief("Value %s out of range, try values between 0 and 100\n", optarg);
        }
        break;
      case 'V':
        version();
        exit(0);
        break;
      case 'h':			/* need help */
      default:
        usage(argv[0]);
        exit(1);
    }
  }

  /* intialize singals' hendler */
  mpc123_initsigh();

  /* set default ao_driver and ao_device if none were
   * specified on the command line
   */
  if( !options.ao_driver ){
    options.ao_driver="oss";
  }
  if( !options.ao_dev ){
    /* if the driver is set, but not the output device, use
     * the default devices */
    mpc123_choose_default_dev_by_driver();
  }

  /* no playlist? and no files on the command line either? too bad! */
  if( !(argv[optind] || options.playlist) ){
    usage(argv[0]);
    exit(1);
  }

  say(1, COPYRIGHT_NOTICE "\n");

  /* dump options */
  debugf("opts.shuffle=%d\n"
         "  opts.random=%d\n"
         "  opts.verbosity=%d\n"
         "  opts.foutput=\"%s\"\n"
         "  opts.playlist=\"%s\"\n"
         "  opts.ao_driver=\"%s\"\n"
         "  opts.ao_dev=\"%s\"\n"
         "  opts.volume=%f",
         options.shuffle, options.random, options.verbosity,
         options.foutput, options.playlist, options.ao_driver,
         options.ao_dev, options.volume);

  /* optind? :) */
  debugf("argv[optind (%d)]=\"%s\"", optind, argv[optind]);

  /* prepare the playlist */
  if(options.playlist){
    if((ret=populate_playlist_from_file(&pl, options.playlist)))
      dief("Error populating playlist from file \"%s\": error %d\n",
           options.playlist, ret);
  }else{
    if((ret=populate_playlist_from_argv(&pl, argv, optind, argc-optind)))
      dief("Error populating playlist from command line: error %d\n",
           ret);
  }

  do_play_playlist(&pl);
  free_playlist(&pl);
  return 0;
}

/* vim:ft=c:tw=78:ts=2:et:cin:
 */
